jsonschema-core.xml

```diff
@@ -1483,63 +1566,97 @@
+                                        https://example.com/root.json#/$defs/B
+                                    </t>
+                                </list>
+                            </t>
+                            <t hangText="#/$defs/B/$defs/X">
+                                <list style="hanging">
+                                    <t hangText="base URI">https://example.com/other.json</t>
+                                    <t hangText="canonical URI with plain fragment">
+                                        https://example.com/other.json#bar
+                                    </t>
+                                    <t hangText="canonical URI with pointer fragment">
+                                        https://example.com/other.json#/$defs/X
+                                    </t>
+                                    <t hangText="Fragment relative to root.json (undefined behavior)">
+                                        https://example.com/root.json#/$defs/B/$defs/X
+                                    </t>
+                                </list>
+                            </t>
+                            <t hangText="#/$defs/B/$defs/Y">
+                                <list style="hanging">
+                                    <t hangText="base URI">https://example.com/t/inner.json</t>
+                                    <t hangText="canonical URI with plain fragment">
+                                        https://example.com/t/inner.json#bar
+                                    </t>
+                                    <t hangText="canonical URI with pointer fragment">
+                                        https://example.com/t/inner.json#
+                                    </t>
+                                    <t hangText="Fragment relative to other.json (undefined behavior)">
+                                        https://example.com/other.json#/$defs/Y
+                                    </t>
+                                    <t hangText="Fragment relative to root.json (undefined behavior)">
+                                        https://example.com/root.json#/$defs/B/$defs/Y
+                                    </t>
+                                </list>
+                            </t>
+                            <t hangText="#/$defs/C">
+                                <list style="hanging">
+                                    <t hangText="base URI">
+                                        urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f
+                                    </t>
+                                    <t hangText="canonical URI with pointer fragment">
+                                        urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f#
+                                    </t>
+                                    <t hangText="Fragment relative to root.json (undefined behavior)">
```

Comment:
This one's not undefined

Reply:
Did you mean to comment on this one?  It says "canonical URI with pointer fragment", as it has the `#` tacked on to the `urn:...`.  The next one relative to `root.json` is undefined.

Reply:
@jdesrosiers oh whoops, the sidescroll bar obscure the line for me :-/

`https://example.com/root.json#/$defs/C` is indeed undefined.  The `urn:...` is an absolute URI and the canonical URI of that schema.  URIs with fragments are only canonical if they use the canonical base.

`https://example.com/root.json#/$defs/C` is effectively the _retrieval URI_ of that subschema.  It's how you get there, but if you are writing an application or another schema that `$ref`s this, you SHOULD use `urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f` or `urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f#` (which are equivalent due to JSON Schema's fragment rules).

Reply:
The more practical reason that they are undefined is that if you changed the example so it looks like this:

```JSON
{
    "$id": "https://example.com/root.json",
    "$defs": {
        "A": { "$anchor": "foo" },
        "B": {
            "$ref": "other.json",
        },
        "C": {
            "$ref": "urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f"
        }
    }
}
```

Then `https://example.com/root.json#/$defs/C` no longer points to the same thing as `urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f`.  It points the `$ref` object.

So if you switch from embedded to reference, that `#/$defs/C` pointer does ***not*** become _invalid_, but it does point to a different thing.  That ultimately behaves the same, but is still not identical.

So I decided that was the criteria for "undefined behavior".  You can evaluate the pointer and it's still valid, but you might be surprised at what you get.

I am open to suggestion on how to present that better, I agree it is puzzling because "undefined" kind of implies that evaluating the JSON Pointer might not be defined.  That's not really what I meant, I mean that it might not always give you the same thing for two schema arrangements that are logically the same.

Reply:
@jdesrosiers OK I ended up reworking the wording a lot in one of the commits I just added.  See what you think of that.  I'm not thrilled with it, but I think it's at least in the right direction?  Further suggestions welcome.

Reply:
> Then `https://example.com/root.json#/$defs/C` no longer points to the same thing as `urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f`. It points the `$ref` object.

This is contradictory to the way I understood it. To me, the following two are equivalent.

`{ "$ref": "urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f" }`
`{ "$id": "urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f", "type": "string" }`

Pointing to the first results in the second. I thought that's what was meant by an `$id` being equivalent to an embedded `$ref`. At least that's what I meant by it when I introduced the concept.

But, I realize that with the changes to `$ref`, that's not even possible anymore. A reference is no longer `{ "$ref": "..." }`, it's the value of the value of a keyword called `$ref`. That changes the dynamics significantly. Now I understand why you proposed that the following should be equivalent.

`{ "$ref": "urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f" }`
`{ "$ref": { "$id": "urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f", "type": "string" } }`

Given the new definition of `$ref`, I think this is the only way the embedding concept makes sense.

Reply:
@jdesrosiers yeah, allowing adjacent keywords changes things.

It also opens up a substantial use case of annotations that can layer semantics (such as OO inheritance or composition) on top of references, which couldn't be done in the old system, and also couldn't be done if we eliminated `$ref` and just used URI strings in place of schema objects.  These use cases specifically rely on being adjacent to something that is recognizably a reference.